lib/ptracer.h

```diff
@@ -0,0 +1,471 @@
+                        sprintf(buf, "    %s:%d: in function '%s'\n",
+                            stack[index].shared.details->filename,
+                            stack[index].line, stack[index].shared.details->fn_name);
+                        DBG_PRINT();
+                    }
+
+                    /* 'check' idx is guaranteed to be a Lua interface frame.
+                       Which is basically our 'stack' index at this point. So,
+                       we simply ignore the Lua interface frame. */
+                    index--;
+
+                    /* We are done. */
+                    lua_settop(L, top);
+                    continue;
+                }
+            }
+
+            /* Then it's an untracked C frame. */
+            if(_pgf_name(L))
+                lua_pushfstring(L, "%s", lua_tostring(L, -1));
+            else lua_pushliteral(L, "<?>");
+
+            sprintf(buf, "    C: in function '%s'\n", lua_tostring(L, -1));
+            DBG_PRINT();
+        } else {
+            /* It's a Lua frame. */
+
+            /* Do we have a name? */
+            if(*ar.namewhat != '\0')
+                lua_pushfstring(L, "function '%s'", ar.name);
+            /* Is it the main chunk? */
+            else if(*ar.what == 'm')
+                lua_pushliteral(L, "<main>");
+            /* Can we deduce the name from the global table? */
+            else if(_pgf_name(L))
+                lua_pushfstring(L, "function '%s'", lua_tostring(L, -1));
+            else lua_pushliteral(L, "function '<?>'");
+
+            sprintf(buf, "    %s:%d: in %s\n", ar.short_src,
+                ar.currentline, lua_tostring(L, -1));
+            DBG_PRINT();
+        }
+
+        lua_settop(L, top);
+    }
+
+    return 0;
+}
+#undef DBG_PRINT
+
+/* When we encounter an error, `pallene_tracer_frameexit()` may not
+   get called. Therefore, the stack will get corrupted if the previous
+   call-frames are not removed. The finalizer function makes sure that
+   it does not happen. It's guardian angel. */
+/* The finalizer function will be called from a to-be-closed value (since
+   Lua 5.4). If you are using Lua version prior 5.4, you are outta luck. */
+int pallene_tracer_finalizer(lua_State *L) {
+    /* Get the userdata. */
+    pt_cont_t *cont = (pt_cont_t *) lua_touserdata(L, lua_upvalueindex(1));
+
+    /* Remove all the frames until last Lua frame. */
+    int idx = cont->count - 1;
+    while(cont->stack[idx].type != PALLENE_TRACER_FRAME_TYPE_LUA)
+        idx--;
+
+    /* Remove the Lua frame as well. */
+    cont->count = idx;
+
+    return 0;
+}
+
+/* Initializes the Pallene Tracer. The initialization refers to creating the stack
+   if not created, preparing the traceback fn and finalizers. */
+/* NOTE: Pushes the finalizer object to the stack. The object has to be closed
+   everytime you are in a Lua C funciton using `lua_toclose(L, idx)` function. */
+pt_cont_t *pallene_tracer_init(lua_State *L) {
```

Comment:
Can we move this higher up in the  file? It makes sense to read this before we read the traceback.